src/atldld/sync.py

```diff
@@ -260,59 +227,190 @@ def download_parallel_dataset(
+
+        self.metadata: Dict[str, Any] = {}
+        # populated by calling `fetch_metadata`
+
+    def __len__(self) -> int:
+        """Return the number of images in the dataset."""
+        if "images" not in self.metadata:
+            raise ValueError("The metadata is empty. Please run `fetch_metadata`")
+        return len(self.metadata["images"])
+
+    def fetch_metadata(
+        self,
+        force_redownload: bool = False,
+    ) -> None:
+        """Fetch metadata of the dataset.
+
+        This function performs the following steps:
+        1. Get metadata for the entire dataset (e.g. `affine_3d`)
+        2. Get metadata for all images inside of the dataset (e.g. `affine_2d`)
+
+        Parameters
+        ----------
+        force_redownload
+            If yes, force to redownload the metadata. Otherwise, if
+            the metadata have been computed once, they are not computed again.
+        """
+        if self.metadata and not force_redownload:
+            return
+
+        # Initialize metadata
+        metadata = {}
+
+        # Prepare query parameters
+        parameters_dataset = RMAParameters(
+            model="SectionDataSet",
+            criteria={
+                "id": self.dataset_id,
+            },
+            include=["alignment3d"],
         )
+        parameters_images = RMAParameters(
+            model="SectionImage",
+            criteria={
+                "data_set_id": self.dataset_id,
+            },
+            include=["alignment2d"],
+        )
+        # Query the API
+        r_dataset = rma_all(parameters_dataset)[0]
+        r_images = rma_all(parameters_images)
+
+        # Extract relevant information
+        def extract_template(data, name):
+            template = np.array(AFFINE_TEMPLATES[name])
+            key = "alignment2d" if name in {"tsv", "tvs"} else "alignment3d"
+            return np.vectorize(lambda x: data[key][x])(template)
+
+        metadata["dataset"] = {
+            "id": r_dataset["id"],
+            "affine_tvr": extract_template(r_dataset, "tvr"),
+            "affine_trv": extract_template(r_dataset, "trv"),
+            "plane_of_section_id": r_dataset["plane_of_section_id"],
+            "section_thickness": r_dataset["section_thickness"],
+        }
+
+        images = []
+        for r_image in r_images:
+            images.append(
+                {
+                    "id": r_image["id"],
+                    "affine_tsv": extract_template(r_image, "tsv"),
+                    "affine_tvs": extract_template(r_image, "tvs"),
+                    "section_number": r_image["section_number"],
+                }
+            )
 
-        img = get_image(image_id, downsample=downsample_img)
-
-        if not include_expression:
-            yield image_id, slice_ref_coordinate, img, df
+        images.sort(key=lambda x: -x["section_number"])  # type: ignore
+        metadata["images"] = images  # type: ignore[assignment]
```

Comment:
I think one can avoid both type ignores like this (I haven't tried though):
1. In the lambda cast explicitly to `int`:
```python
images.sort(key=lambda x: -int(x["section_number"]))
```
2. Instead of defining the `metadata` dict and than later assigning `metadata["images"]` define `metadata` later, when `images` already exists:
  ```python
  metadata = {
      "id": ...,
      "images": ...,
      ...
  }
  ```

Reply:
It seems it is working fine (11995ac). Thanks for the suggestions!